server/src/main/java/org/apache/iotdb/db/metadata/mtree/traverser/Traverser.java

```diff
@@ -221,27 +235,77 @@ protected void processNameMatch(IMNode node, int idx, int level) throws Metadata
     }
 
     Template upperTemplate = node.getUpperTemplate();
-    IMeasurementSchema targetSchema = upperTemplate.getSchemaMap().get(targetName);
-    if (targetSchema != null) {
-      traverse(
-          MeasurementMNode.getMeasurementMNode(
-              node.getAsEntityMNode(), targetSchema.getMeasurementId(), targetSchema, null),
-          idx + 1,
-          level + 1);
+
+    IMNode targetNode = upperTemplate.getDirectNode(targetName);
+    if (targetNode != null) {
+      traverseContext.push(node);
+      traverse(targetNode, idx + 1, level + 1);
+      traverseContext.pop();
     }
 
     if (multiLevelWildcard) {
-      for (IMeasurementSchema schema : upperTemplate.getSchemaMap().values()) {
-        traverse(
-            MeasurementMNode.getMeasurementMNode(
-                node.getAsEntityMNode(), schema.getMeasurementId(), schema, null),
-            idx,
-            level + 1);
+      for (IMNode child : upperTemplate.getDirectNodes()) {
+        traverseContext.push(node);
+        traverse(child, idx, level + 1);
+        traverseContext.pop();
       }
     }
   }
 
   public void setPrefixMatch(boolean isPrefixMatch) {
     this.isPrefixMatch = isPrefixMatch;
   }
+
+  /**
+   * @param currentNode the node need to get the full path of
+   * @return full path from traverse start node to the current node
+   */
+  protected PartialPath getPivotPartialPath(IMNode currentNode) throws IllegalPathException {
+    Iterator<IMNode> nodes = traverseContext.descendingIterator();
+    StringBuilder builder = new StringBuilder(nodes.next().getName());
+    while (nodes.hasNext()) {
+      builder.append(TsFileConstant.PATH_SEPARATOR);
+      builder.append(nodes.next().getName());
+    }
+    if (builder.length() != 0) {
+      builder.append(TsFileConstant.PATH_SEPARATOR);
+    }
+    builder.append(currentNode.getName());
+    return new PartialPath(builder.toString());
+  }
+
+  protected MeasurementPath getPivotMeasurementPathInTraverse(IMeasurementMNode currentNode)
+      throws MetadataException {
+    IMNode par = traverseContext.peek();
+    if (!(par instanceof IEntityMNode)) {
+      throw new MetadataException("Measurement not under entity: " + currentNode.getName());
+    }
```

Comment:
Is this necessary? How could the parent MNode of current MeasurementMNode not be Entity?

Reply:
removed now